File: cmd/notice.go
package cmd

import (
	"fmt"
	"log"
	"time"

	"github.com/bwmarrin/discordgo"
	"kuics.ac.kr/discordbot/internal/ctfapi"
)

func Notice(s *discordgo.Session, i *discordgo.InteractionCreate, id string) {
	perm, _ := s.UserChannelPermissions(i.Member.User.ID, i.ChannelID)

	if perm&discordgo.PermissionAdministrator == discordgo.PermissionAdministrator {
		ctfInfo := ctfapi.GetDetailById(id)

		err := s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
			Type: discordgo.InteractionResponseDeferredChannelMessageWithSource,
		})
		if err != nil {
			log.Fatalf("Failed to ACK: %v", err)
		}

		role, err := s.GuildRoleCreate(i.GuildID)
		if err != nil {
			log.Fatalf("Failed to create role: %v", err)
		}

		_, err = s.GuildRoleEdit(i.GuildID, role.ID, ctfInfo.Summary.Name, 0, false, 0, true)
		if err != nil {
			log.Fatalf("Failed to edit role: %v", err)
		}

		_, err = s.GuildChannelCreateComplex(i.GuildID, discordgo.GuildChannelCreateData{
			Name:  ctfInfo.Summary.Name,
			Type:  discordgo.ChannelTypeGuildText,
			Topic: ctfInfo.Url,
			PermissionOverwrites: []*discordgo.PermissionOverwrite{
				{
					ID:    role.ID,
					Type:  discordgo.PermissionOverwriteTypeRole,
					Allow: discordgo.PermissionAll,
				},
			},
		})
		if err != nil {
			log.Fatalf("Failed to create text channel: %v", err)
		}

		channel_voice, err := s.GuildChannelCreateComplex(i.GuildID, discordgo.GuildChannelCreateData{
			Name: ctfInfo.Summary.Name,
			Type: discordgo.ChannelTypeGuildVoice,
			PermissionOverwrites: []*discordgo.PermissionOverwrite{
				{
					ID:    role.ID,
					Type:  discordgo.PermissionOverwriteTypeRole,
					Allow: discordgo.PermissionAll,
				},
			},
		})
		if err != nil {
			log.Fatalf("Failed to create voice channel: %v", err)
		}

		start, _ := time.Parse(time.RFC3339, ctfInfo.Summary.Start)
		finish, _ := time.Parse(time.RFC3339, ctfInfo.Summary.Finish)
		_, err = s.GuildScheduledEventCreate(i.GuildID, &discordgo.GuildScheduledEventParams{
			ChannelID:          channel_voice.ID,
			Name:               ctfInfo.Summary.Name,
			Description:        ctfInfo.Description,
			ScheduledStartTime: &start,
			ScheduledEndTime:   &finish,
			PrivacyLevel:       discordgo.GuildScheduledEventPrivacyLevelGuildOnly,
			Status:             discordgo.GuildScheduledEventStatusActive,
			EntityType:         discordgo.GuildScheduledEventEntityTypeVoice,
		})
		if err != nil {
			log.Fatalf("Failed to create event: %v", err)
		}

		_, err = s.InteractionResponseEdit(s.State.User.ID, i.Interaction, &discordgo.WebhookEdit{
			Components: []discordgo.MessageComponent{
				discordgo.ActionsRow{
					Components: []discordgo.MessageComponent{
						discordgo.Button{
							Label:    "참가신청",
							Style:    discordgo.PrimaryButton,
							CustomID: "apply",
						},
					},
				},
			},
			Embeds: []*discordgo.MessageEmbed{
				{
					Title:       ctfInfo.Summary.Name,
					Description: ctfInfo.Description,
					URL:         ctfInfo.Url,
					Thumbnail: &discordgo.MessageEmbedThumbnail{
						URL: ctfInfo.Logo,
					},
					Fields: []*discordgo.MessageEmbedField{
						{
							Name:  "Format",
							Value: ctfInfo.Format,
						},
						{
							Name:  "Weight",
							Value: fmt.Sprintf("%.2f", ctfInfo.Summary.Weight),
						},
						{
							Name:  "Duration",
							Value: fmt.Sprintf("%s ~ %s\n%d days %d hours", ctfInfo.Summary.Start, ctfInfo.Summary.Finish, ctfInfo.Duration.Days, ctfInfo.Duration.Hours),
						},
					},
					Footer: &discordgo.MessageEmbedFooter{
						Text: role.ID,
					},
				},
			},
		})
		if err != nil {
			log.Fatalf("Failed to send message: %v", err)
		}
	} else {
		log.Println("Permission denied: " + i.Member.User.Username)
	}
}

File: src/handlers/notice.ts
import { APIActionRowComponent, APIButtonComponentWithCustomId, ButtonInteraction, ButtonStyle, CacheType, ChannelType, ComponentType, GuildScheduledEventEntityType, GuildScheduledEventPrivacyLevel, OverwriteType, PermissionFlagsBits } from "discord.js";
import fetch from "node-fetch";
import ctfDetail from "../api/ctfDetail";

export default async (interaction: ButtonInteraction<CacheType>) => {
  await interaction.deferReply();

  const urlSplited = interaction.message.embeds[0].url.split("/")
  const ctfId = urlSplited[urlSplited.length - 1];
  const detail = await ctfDetail(ctfId);

  const role = await interaction.guild.roles.create({
    name: detail.summary.name,
  });
  await interaction.guild.channels.create({
    name: detail.summary.name,
    type: ChannelType.GuildText,
    topic: detail.url,
    permissionOverwrites: [
      {
        id: role.id,
        type: OverwriteType.Role,
        allow: 
          PermissionFlagsBits.AddReactions |
          PermissionFlagsBits.ViewChannel |
          PermissionFlagsBits.SendMessages |
          PermissionFlagsBits.EmbedLinks |
          PermissionFlagsBits.AttachFiles |
          PermissionFlagsBits.ReadMessageHistory |
          PermissionFlagsBits.CreatePublicThreads |
          PermissionFlagsBits.UseExternalStickers |
          PermissionFlagsBits.SendMessagesInThreads |
          PermissionFlagsBits.UseEmbeddedActivities
      },
    ],
  });
  const voiceChannel = await interaction.guild.channels.create({
    name: detail.summary.name,
    type: ChannelType.GuildVoice,
    permissionOverwrites: [
      {
        id: role.id,
        type: OverwriteType.Role,
        allow: 
          PermissionFlagsBits.ViewChannel |
          PermissionFlagsBits.ReadMessageHistory |
          PermissionFlagsBits.Connect |
          PermissionFlagsBits.Speak |
          PermissionFlagsBits.RequestToSpeak
      },
    ],
  });
  let logoImage: Buffer;
  try {
    logoImage = await (await fetch(detail.logo)).buffer()
  } catch (error) {}

  await interaction.guild.scheduledEvents.create({
    name: detail.summary.name,
    scheduledStartTime: detail.startRaw,
    scheduledEndTime: detail.finishRaw,
    privacyLevel: GuildScheduledEventPrivacyLevel.GuildOnly,
    entityType: GuildScheduledEventEntityType.Voice,
    description: detail.description,
    channel: voiceChannel,
    image: logoImage,
  })
  
  const button: APIActionRowComponent<APIButtonComponentWithCustomId> = {
    components: [
      {
        label: "참가신청",
        style: ButtonStyle.Primary,
        custom_id: "apply",
        type: ComponentType.Button,
      },
    ],
    type: ComponentType.ActionRow,
  };
  await interaction.editReply({
    components: [button],
    embeds: [
      {
        title: detail.summary.name,
        description: detail.description,
        url: detail.url,
        thumbnail: {
          url: detail.logo,
        },
        fields: [
          {
            name: "Format",
            value: detail.summary.format,
          },
          {
            name: "Weight",
            value: detail.summary.weight.toString(),
          },
          {
            name: "Duration",
            value: `${detail.summary.start} ~ ${detail.summary.finish}\n${detail.duration.days}일 ${detail.duration.hours}시간`,
          },
        ],
        footer: {
          text: role.id,
        },
      },
    ],
  });
}
